gprofiler/profilers/java.py

```diff
@@ -259,11 +287,86 @@ def make_application_metadata(self, process: Process) -> Dict[str, Any]:
         # that loads other libs.
         libjvm_elfid = get_mapped_dso_elf_id(process, "/libjvm")
 
-        metadata = {"java_version": version, "libjvm_elfid": libjvm_elfid}
+        metadata = {
+            "java_version": version,
+            "libjvm_elfid": libjvm_elfid,
+            "jvm_flags": self.get_jvm_flags_serialized(process),
+        }
 
         metadata.update(super().make_application_metadata(process))
         return metadata
 
+    def get_jvm_flags_serialized(self, process: Process) -> Optional[List[Dict]]:
+        jvm_flags = self.get_jvm_flags(process)
+        if jvm_flags is None:
+            return None
+
+        return [flag.to_dict() for flag in sorted(jvm_flags, key=lambda flag: flag.name)]
+
+    def get_jvm_flags(self, process: Process) -> Optional[Iterable[JvmFlag]]:
+        jvm_raw_flags: Optional[Iterable[JvmFlag]] = self.get_jvm_flags_raw(process)
+
+        if jvm_raw_flags is None:
+            return None
+
+        jvm_raw_flags = filter(self.filter_jvm_flag, jvm_raw_flags)
+
+        if self.jvm_flags_to_collect == JavaFlagCollectionOptions.ALL:
+            return jvm_raw_flags
+        elif self.jvm_flags_to_collect == JavaFlagCollectionOptions.DEFAULT:
+            return filter(self.default_collection_filter_jvm_flag, jvm_raw_flags)
+        elif isinstance(self.jvm_flags_to_collect, list):
+            requested_flags = [flag for flag in jvm_raw_flags if flag.name in self.jvm_flags_to_collect]
+            not_found_flags = set(self.jvm_flags_to_collect) - {flag.name for flag in requested_flags}
+
+            # log if the set is not empty
+            if not_found_flags:
+                logger.error("Flags not found in jvm flags:", not_found_flags=not_found_flags)
+
+            return requested_flags
+        else:
+            assert False, f"Unrecognized jvm_flags_to_collect value: {self.jvm_flags_to_collect}"
```

Comment:
Uhh less nicer that way, instead you should remove the condition in the last `if`, and `assert` that condition itself.